nvflare/app_common/widgets/log_streaming.py

```diff
@@ -0,0 +1,122 @@
+# limitations under the License.
+
+import os
+import threading
+from builtins import dict as StreamContext
+
+from nvflare.apis.event_type import EventType
+from nvflare.apis.fl_constant import FLContextKey, ReturnCode, SystemComponents
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.workspace import Workspace
+from nvflare.app_common.streamers.file_streamer import FileStreamer
+from nvflare.widgets.widget import Widget
+
+LOG_STREAM_EVENT_TYPE = "stream_log"
+
+
+class LogConst(object):
+    CLIENT_NAME = "client_name"
+    JOB_ID = "job_id"
+    LOG_DATA = "log_data"
+
+
+class LogSender(Widget):
+    def __init__(self, event_type=EventType.JOB_COMPLETED, should_report_error_log: bool = True):
+        """Sender for analytics data."""
+        super().__init__()
+        self.event_type = event_type
+        self.should_report_error_log = should_report_error_log
+
+    def _stream_error_log_file(self, fl_ctx: FLContext):
+        error_log_contents = None
+        workspace_root = fl_ctx.get_prop(FLContextKey.WORKSPACE_ROOT)
+        client_name = fl_ctx.get_prop(FLContextKey.CLIENT_NAME)
+        job_id = fl_ctx.get_prop(FLContextKey.CURRENT_JOB_ID)
+        workspace_object = Workspace(root_dir=workspace_root, site_name=client_name)
+        error_log_path = workspace_object.get_app_error_log_file_path(job_id=job_id)
+        if os.path.exists(error_log_path):
+            with open(error_log_path, "r") as f:
+                error_log_contents = f.read()
+        if error_log_contents:
+            FileStreamer.stream_file(
+                channel="error_logs",
+                topic=LOG_STREAM_EVENT_TYPE,
+                stream_ctx={LogConst.CLIENT_NAME: client_name, LogConst.JOB_ID: job_id},
+                targets=["server"],
+                file_name=error_log_path,
+                fl_ctx=fl_ctx,
+            )
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == self.event_type:
+            if self.should_report_error_log:
+                t = threading.Thread(target=self._stream_error_log_file, args=(fl_ctx,), daemon=True)
+                t.start()
+                client_name = fl_ctx.get_prop(FLContextKey.CLIENT_NAME)
+                job_id = fl_ctx.get_prop(FLContextKey.CURRENT_JOB_ID)
+                self.log_info(fl_ctx, f"Started streaming error log file for {client_name} for job {job_id}")
+
+
+class LogReceiver(Widget):
+    def __init__(self, log_channels=None):
+        """Receives log data.
+
+        By default, it receives error logs with the channel error_logs to save logs of type ERRORLOG. If adding
+        additional log types, make sure nvflare.apis.storage.ComponentPrefixes has the corresponding log type.
+
+        Args:
+            log_channels: dict of channel to log type mapping
+        """
+        super().__init__()
+        if log_channels is None:
+            log_channels = {"error_logs": "ERRORLOG"}
```

Comment:
Please define some symbols for the channel name and log type, instead of hard-coding.